### PR TITLE
Support basic trigonometric functions

### LINKED ISSUE #93: Suport basic trigonometric functions
## Details:
Hours Spent: 0

## Acceptance Criteria
- [x] Approx pi to arbitrary accuracy is available
- [x] Approx sine and cosine are available and work correctly

## Closing Notes

Overview of changes can be found in PR #97

### PR DESCRIPTION
**Pull request by michalkonecny**
# Issues
Closes #93 

# Summary
- Added robust pi
- Added robust sin, cos, tan
- Fixed differentiation of constants and other variables than x

# Tests
- tests for some properties of Approx pi, sine and cosine
- evaluation tests for sine,  cosine and tan

### CODE REVIEW

File: src/Expression/Evaluator.purs [hunk 1]
```diff
@@ -9,27 +9,27 @@ import Expression.Error (Expect, evaluationError, multipleErrors, throw, unknown
...
+  ExpressionVariable name -> case lookup (presetConstants <> variableMap) name of
     Just value -> pure value
     _ -> unknownValue name
   ExpressionBinary operation leftExpression rightExpression -> roughEvaluateBinaryOperation operation variableMap leftExpression rightExpression
   ExpressionUnary operation expression -> roughEvaluateUnaryOperation operation variableMap expression
   ExpressionLet name expression parentExpression -> case roughEvaluate variableMap expression of
     Right value -> roughEvaluate (variableMap <> [ (Tuple name value) ]) parentExpression
     Left error -> Left error
+  where
+  presetConstants = [ (Tuple "pi" pi) ]
```
Comment (RelativeForce) [inline src/Expression/Evaluator.purs:34]: How come you removed `e` from the rough plot?

File: src/IntervalArith/Approx/Approx.purs [hunk 2]
```diff
@@ -633,4 +707,23 @@ increasingPartialFunctionViaBounds :: (Approx -> Maybe Approx) -> (Approx -> May
...
   resultU <- f $ upperA a
-  pure $ resultL `unionA` resultU
\ No newline at end of file
+  pure $ resultL `unionA` resultU
+
+{-|
+Second argument is noise to be added to first argument. Used to allow for the
+error term when truncating a series.
+-}
+fudge :: Approx -> Approx -> Approx
```
Comment (RelativeForce) [inline src/IntervalArith/Approx/Approx.purs:716]: Is this a temporary function name? Can it be changed to something more descriptive?

File: src/IntervalArith/Approx/Pi.purs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+The main idea is to balance the computations so that more operations are
+performed with values of similar size. Using the underlying fast
+multiplication algorithms this will give performance benefits.
+
+The algorithm parallelises well. However, a final division is needed at the
+end to compute /T\/BQ/ which amount to a substantial portion of the
+computation time.
+-}
+abpq ::
```
Comment (RelativeForce) [inline src/IntervalArith/Approx/Pi.purs:39]: Could this function name be more descriptive?

File: src/Misc/Tuple.purs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module Misc.Tuple where
```
Comment (RelativeForce) [inline src/Misc/Tuple.purs]: Renamed to `Tuple22`

**Review by RelativeForce — COMMENTED** [review]
Looks good! Just a few minor suggestions

File: src/Expression/Evaluator.purs (hunk 1 above)
Comment (michalkonecny) [reply src/Expression/Evaluator.purs:34]: I plan to put it back when we have it also in the robust evaluate.

File: src/IntervalArith/Approx/Approx.purs (hunk 2 above)
Comment (michalkonecny) [reply src/IntervalArith/Approx/Approx.purs:716]: This comes directly from CDAR.  I prefer to keep it in sync with CDAR although some of the names are not very descriptive. Also, in this case it is hard to find a better name.  I think that fudge is a reasonable name for what it does - artificially lose precision.

File: src/Misc/Tuple.purs (hunk 4 above)
Comment (michalkonecny) [reply src/Misc/Tuple.purs]: Thank you, done.

File: src/IntervalArith/Approx/Pi.purs (hunk 3 above)
Comment (michalkonecny) [reply src/IntervalArith/Approx/Pi.purs:39]: A bit like with fudge, I am reluctant to deviate from CDAR although I agree that this name does not help very much work out what the function does.  Fortunately, the comment is quite helpful in pointing us to the research paper from which this comes should we need a full explanation.

**Review by RelativeForce — APPROVED** [review]
Looks alright

**Pull request merged**